### PR TITLE
Gas estimator

### PR DESCRIPTION
**Pull request by Agusx1211**
- Add Estimator package
- Move test runtime from ganache to geth
- Add GasEstimator contracts
- Use new EthKit version with support for deployed bytecode
- Add gasLimits for create wallet and universal deployer

### CODE REVIEW

File: estimator.go [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+	// Compute gas estimation for slices of all transactions
+	// including no transaction execution and all transactions
+
+	nonce, err := txs.Nonce()
+	if err != nil {
+		return 0, err
+	}
+
+	for i := range estimates {
```
Comment (PhABC) [inline estimator.go:313]: Ideally we could also do this in parallel via multicall, but that's a further optimization

File: estimator.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+
+	var success bool
+	var result []byte
+	var gas *big.Int
+
+	if err := ethcoder.AbiDecoder([]string{"bool", "bytes", "uint256"}, resBytes, []interface{}{&success, &result, &gas}); err != nil {
+		return nil, err
+	}
+
+	gas.Add(gas, big.NewInt(int64(e.CalldataCost(call.Data))))
```
[the remark this answers is not in the document]
Comment (Agusx1211) [reply estimator.go:129]: No, because the estimateGas call doesn't return gas usage, it just calls a method of a contract. We use it by calling a method that relays the transaction, this method executes the transaction does the gas metering and returns the result, but it cannot account for neither the calldata cost or the base transaction cost.

File: estimator.go [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+	}
+
+	return gas, nil
+}
+
+func (e *Estimator) AreEOAs(ctx context.Context, provider *ethrpc.Provider, walletConfig WalletConfig) ([]bool, error) {
+	// Get non-eoa signers
+	// required for computing worse case scenario
+	// TODO: Do these calls in parallel
+	isEoa := make([]bool, walletConfig.Signers.Len())
```
[the remark this answers is not in the document]
Comment (Agusx1211) [reply estimator.go:143]: Yep we can! This first version just tries to keep it simple

File: estimator.go (hunk 1 above)
Comment (Agusx1211) [reply estimator.go:313]: Same as before, once we have a ton of tests we can play around with those optimisations

File: transactions.go [hunk 4]
```diff
@@ -362,6 +397,48 @@ func GetWalletNonce(provider *ethrpc.Provider, walletConfig WalletConfig, wallet
...
+	}
+	if logs[0].Topics[0] != NonceChangeEventSig {
+		return false
+	}
+	return true
+}
+
+// prepareTransactionsForEncoding checks the transactions data structure with basic
+// integrity checks
+func prepareTransactionsForEncoding(txns Transactions) (Transactions, error) {
```
Comment (attente) [inline transactions.go:412]: Can you remove this and use `Transactions.EncodedTransactions()` instead? It was renamed on master.

File: transactions.go [hunk 5]
```diff
@@ -362,6 +397,48 @@ func GetWalletNonce(provider *ethrpc.Provider, walletConfig WalletConfig, wallet
...
+			}
+		}
+
+		stxns = append(stxns, txn)
+	}
+
+	return stxns, nil
+}
+
+func EncodeTransactions(txns Transactions) ([]byte, error) {
```
Comment (attente) [inline transactions.go:438]: This function was also removed in favour of `Transactions.Execdata()`.

File: transactions.go [hunk 6]
```diff
@@ -362,6 +397,48 @@ func GetWalletNonce(provider *ethrpc.Provider, walletConfig WalletConfig, wallet
 	return nonceResult, nil
 }
 
+func IsNonceChangedEvent(logs []*types.Log) bool {
```
Comment (attente) [inline transactions.go:400]: This was also removed on master.

File: estimator.go [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+	return willSign, nil
+}
+
+func stubAddress() common.Address {
+	raw := make([]byte, 20)
+	rand.Read(raw)
+	return common.BytesToAddress(raw)
+}
+
+func (e *Estimator) BuildStubSignature(walletConfig WalletConfig, willSign []bool, isEoa []bool) []byte {
```
Comment (attente) [inline estimator.go:207]: Just a thought, but what if instead of building the stub signature in the estimator, we have the wallets themselves be able to construct their own stub signatures since they're likely more aware of what signers they will be using?

Comment (PhABC) [reply estimator.go:207]: Relayer would still need it's own independent logic because the wallet could lie about the using the worst case scenario.

Comment (Agusx1211) [reply estimator.go:207]: Yep exactly, but this may not be a bad idea for the nested signers problem.

The wallet could provide the Config for all nested signers, then the relayer could validate if the provided config matches the configuration of those signers.

But it would add more complexity to the whole thing, I think it's better to keep it simple for now.

Comment (attente) [reply estimator.go:207]: But if the wallet lies, it would be doing it to its own detriment, no? The wallet would lie, saying it expects to sign with a really short signature, then sign with a really long signature, which would fail the relayer's validation check.

File: signature.go [hunk 8]
```diff
@@ -77,7 +77,7 @@ func (s *Signature) Encode() ([]byte, error) {
 		case sigPartTypeDynamic:
 			pack, err = ethcoder.SolidityPack(
 				[]string{"uint8", "uint8", "address", "uint16", "bytes"},
-				[]interface{}{signer.Type, signer.Weight, signer.Address, len(signer.Value), signer.Value},
+				[]interface{}{signer.Type, signer.Weight, signer.Address, big.NewInt(int64(len(signer.Value))), signer.Value},
```
Comment (attente) [inline signature.go:80]: Why is this change necessary?
